accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -616,6 +609,442 @@ It is a warning to use a null aware operator (`?.`, `?[]`, `?..`, `??`, `??=`, o
+reachable paths to the join point and not on another reachable path.  Note that
+it is valid for a `final` variable to have no **initialization inferred type**
+on a path into a join point when the join point is known to be unreachable along
+that path.
+
+```dart
+void test(bool b) {
+  final x; // x is an untyped variable
+  if (b) {
+    x = 3;
+  } // Error, x has not been given an inferred type on all paths
+
+  late final y; // y is a late untyped variable
+  if (b) {
+    y = 3;
+  } // Ok, y has inferred type `int`
+
+  final z;
+  if (b) {
+     z = 3;
+  } else {
+    return;
+  }  // Ok, z has inferred type `int` even though it is unassigned on one path.
+}
+```
+
+At join points, some adjustment to the **initialization inferred type** is done
+to account for nullability.  Informally, a non-`late`, `final` **untyped**
+variable must be assigned a value on every path, but a variable declared `late`
+and/or `var` may be assigned a value on only some of the paths into a join.  In
+the latter case, the type is made nullable if the variable is declared as
+non-`late`, and the variable is treated as implicitly null initialized.
+Variables declared `late` are only made nullable if `Null` is explicitly
+assigned on one more paths into the join, and there is no implicit null
+initialization.  In all cases, the values assigned must all have the same type
+(up to nullability), or type `Null` - otherwise the variable becomes unusable
+after the join.  More precisely, the **initialization inferred type** of a
+variable after a join point is specified as follows.
+
+At a join point in the program, if an **untyped variable** has no
+**initialization inferred type** on any path into the join, then it has no
+**initialization inferred type** after the join.
+
+Otherwise, if an **untyped variable** has an **initialization inferred type** on
+at least one path (reachable or not) into the join, then we consider two cases:
+first that it has an **initialization inferred type** on every reachable path
+into the join (in which case it does not need to be made implicitly nullable),
+or that it has at least one reachable path with no **initialization inferred
+type** (in which case it must be made implicitly nullable if it is non-`late`).
+
+At a join point in the program, if an **untyped variable** has an
+**initialization inferred type** on all of the *reachable* paths into the join,
+then the **initialization inferred type** for the variable after the join point
+is computed as follows:
+  - If the **initialization inferred type** on each path which has one
+    (reachable or not) is the same type `T` (using syntactic equality), then the
+    **initialization inferred type** after the join is `T`.
+  - Otherwise, if the **initialization inferred type** on each path which has
+    one (reachable or not) is one of `T`, `T?` or `Null` for some type `T`
+    (using syntactic equality), then the **initialization inferred type** after
+    the join is `T?`.
```

Comment:
This case would match if the IIT on all paths is `Null`, but in that case, you don't know what T should be:

```dart
var x;
if (c) {
  x = null;
} else {
  x = null;
}
// what is x here?
```

This is a corner where my proposal gives you an answer. It says `x` is in the potentially assigned `Null` type state, which means it can be used as `null` but can still have a later type inferred:

```dart
var x; // Unassigned+Untyped
if (c) {
  x = null; // Assigned+Null
} else {
  x = null; // Assigned+Null
} // Join = Assigned+Null
print(x); // OK, null.
x = 123; // OK, Assigned+Typed(int?)
x = "str"; // Error, wrong type.
```

(Whether the answer my proposal provides is a *good* answer is an open question. :) )

Reply:
As discussed, `Null` is a solution for `T` in `T` or `T?` or `Null`, and so this proposal also gives `Null` as the solution here.